### PR TITLE
P2P part1

### PR DESCRIPTION
**Pull request by asdacap**
Part1 of probably several for p2p codes.
- Basic peer discovery via kademlia.
- Peer pool implementation, which will be used by sync code later.

### CODE REVIEW

**Review by omerfirmak — APPROVED** [review]
I am fine with merging this even after the release.

**Comment by asdacap** [comment]
Lets merge this.

**Comment by omerfirmak** [comment]
Fixed the history, please use rebase only to maintain a linear history. Merge commits makes it realy hard to collaborate on a branch

**Comment by omerfirmak** [comment]
I want to merge this as well, @NethermindEth/nubia-juno please take a look.

File: node/node.go [hunk 1]
```diff
@@ -122,6 +129,17 @@ func New(cfg *Config, version string) (*Node, error) {
 		n.services = append(n.services, grpc.NewServer(n.cfg.GRPCPort, n.version, n.db, n.log))
 	}
 
+	if cfg.P2P {
+		privKeyStr, _ := os.LookupEnv("P2P_PRIVATE_KEY")
```
Comment (joshklop) [inline node/node.go:133]: Is it possible to pass this in as a flag? Not sure if that's worse from a security point of view.

**Review by joshklop — APPROVED** [review]

File: node/node.go (hunk 1 above)
Comment (omerfirmak) [reply node/node.go:133]: We can add a flag for using keystore files in the future.

**Review by IronGauntlets — APPROVED** [review]

**Pull request merged**